src/vsphere_cpi/lib/cloud/vsphere/attach_tag_to_vm.rb

```diff
@@ -0,0 +1,134 @@
+        end
+        nil
+      end
+
+      def vm_association?(target_category_id, config_category)
+        category_information = tagging_category_api_instance.get(target_category_id)
+        category_assoc_types = category_information.value.associable_types
+        return true if category_assoc_types.empty? || category_assoc_types.include?("VirtualMachine")
+        logger.warn("Tag category '#{config_category}' is not associated with object type: 'Virtual Machine', skip attaching this category to vm." )
+        false
+      end
+
+      def single_category_cardinality?( target_category_id )
+        category_information= tagging_category_api_instance.get(target_category_id)
+        return true if category_information.value.cardinality == "SINGLE"
+        false
+      end
+
+      def create_tag_hash(vm_config)
+        vm_config_tags =  vm_config.vm_type.tags
+        tag_hash = Hash.new
+        vm_config_tags.each do |vm_config_tag|
+          if vm_config_tag["tag"].nil?
+            logger.warn("Empty tag found in category '#{vm_config_tag["category"]}', skip attaching this tag." )
+            next
+          end
+          tag_hash[vm_config_tag["category"]] ||= []
+          tag_hash[vm_config_tag["category"]] << vm_config_tag["tag"]
+        end
+        tag_hash
+      end
+
+      def attach_single_tag(created_vm, config_category, tag_array, category_tags)
+        target_tag = verify_tag_id(tag_array[0], category_tags)
+        if target_tag.nil?
+          logger.warn("Tag '#{config_category}:#{tag_array[0]}' can not be found, skip attaching this tag to vm." )
+          return
+        end
+
+        tag_assoc_info.object_id = VSphereAutomation::CIS::VapiStdDynamicID.new
+        tag_assoc_info.object_id.id = created_vm.mob_id
+        tag_assoc_info.object_id.type = 'VirtualMachine'
+        tag_assoc_api.attach(target_tag,tag_assoc_info)
+      end
+
+      def attach_multi_tags(created_vm, config_category, target_category_id, tag_array, category_tags)
+        single_cardinality = single_category_cardinality?(target_category_id)
+        if single_cardinality
+          logger.warn("Tag category '#{config_category}' does not support 'Multiple Cardinality', skip tagging this category to vm." )
+          return
+        end
+        tag_ids = []
+        tag_array.each do |tag_name|
+          target_tag = verify_tag_id(tag_name, category_tags)
+          tag_ids << target_tag unless target_tag.nil?
+        end
+        multi_tag_assoc_info.object_id=VSphereAutomation::CIS::VapiStdDynamicID.new
+        multi_tag_assoc_info.object_id.id=created_vm.mob_id
+        multi_tag_assoc_info.object_id.type='VirtualMachine'
+        multi_tag_assoc_info.tag_ids=tag_ids
+        tag_assoc_api.attach_multiple_tags_to_object(multi_tag_assoc_info)
+      end
+
+      def attach_tags(created_vm, vm_config)
+        category_id_list_on_host = tagging_category_api_instance.list
```

Comment:
`category_id_list_on_host` what do you mean by host here? Is it targeted vCenter? You can change name to `all-category_ids`

Reply:
Done. Use `category_ids` instead, there is only one  category id groups( array ) in codebase.